### PR TITLE
Add bug report load/save for Devtools.

### PR DESCRIPTION
**Pull request by yuangu-google**

### CODE REVIEW

File: devtools/src/arcs-bug-report.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    }, false);
+  }
+
+  onRawMessageBundle(messages) {
+    this.messages = this.messages.concat(messages);
+  }
+
+  _saveBugReport() {
+    var zip = new JSZip();
+    var file = zip.file("bugreport.txt", JSON.stringify(this.messages));
```
Comment (piotrswigon) [inline devtools/src/arcs-bug-report.js]: It doesn't appear like JSZip is compressing the file. The zip has the same size (actually, is slightly larger) that the txt file. But if I unzip it and zip it locally on my machine I went from 2.1MB to 216KB...

I've looked around and found this: https://github.com/Stuk/jszip/issues/406

File: devtools/src/arcs-bug-report.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    const fileElem = this.$.fileElem;
+    this.$.fileSelect.addEventListener("click", function (e) {
+      if (fileElem) {
+        fileElem.click();
+      }
+    }, false);
+  }
+
+  onRawMessageBundle(messages) {
+    this.messages = this.messages.concat(messages);
```
Comment (piotrswigon) [inline devtools/src/arcs-bug-report.js]: Or: `this.messages.push(...messages);`
It will create fewer arrays, which I guess is better.

File: devtools/src/arcs-bug-report.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    <style include="shared-styles">
+      :host {
+        display: flex;
+        flex-direction: column;
+      }
+    </style>
+    <div class="wrapper">
+      <input type="file" id="fileElem" style="display:none" on-change="_loadBugReport">
+      <button id="fileSelect">Load Bug Report</button>
+      <button type="button" on-click="_saveBugReport">Save Bug Report</button>
```
Comment (piotrswigon) [inline devtools/src/arcs-bug-report.js]: Maybe `<iron-icon title="Download Bug Report" icon="bug-report" on-click="fileBugReport"></iron-icon>` ?

It would draw a "bug" icon:
Screen Shot 2019-09-24 at 10 22 18 am

File: devtools/src/arcs-bug-report.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        a.click();
+      });
+  }
+
+  _loadBugReport(event) {
+    let files = event.target.files;
+    for (var i = 0; i < files.length; i++) {
+      JSZip.loadAsync(files[i]).then(
+        zip => this._decompress(zip),
+        e => alert("Error reading " + f.name + ": " + e.message));
```
Comment (piotrswigon) [inline devtools/src/arcs-bug-report.js]: there's no `f`, `files[i]` instead?
or
`for (f of files) {`

File: devtools/src/arcs-bug-report.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+  _decompress(zip) {
+    zip.forEach(
+      relativePath => zip.file(relativePath).async("text").then(
+        txt => this._handleLog(txt)));
+  }
+
+  _handleLog(txt) {
+    let messages = JSON.parse(txt);
+    this.emitFilteredMessages(messages);
```
Comment (piotrswigon) [inline devtools/src/arcs-bug-report.js]: You may have noticed that it doesn't fully replicate the state the Arcs Explorer was in. (e.g. all the arcs are shown together in the overview, you cannot select between arcs etc.

This is because you are replaying the messages as these "filtered messages", which should only pertain to a single arc. arcs-selector is the component that is responsible for switching between arcs and replaying the messages, but it itself listens to messages which are _not_ filtered.

To sum up, you should not emit the messages as filtered messages, you should emit them as they are and let the webapp deal with them as it would. The easiest way to do it would be:

`document.dispatchEvent(new CustomEvent('raw-messages', {detail: messages}));`

A slightly nicer way would be to have a concept of "replayed connection" in devtools.js, where all the connection types are listed. I'll let you explore that if you're interested.

File: devtools/src/arcs-bug-report.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    return html`
+    <style include="shared-styles">
+      :host {
+        display: flex;
+        flex-direction: column;
+      }
+    </style>
+    <div class="wrapper">
+      <input type="file" id="fileElem" style="display:none" on-change="_loadBugReport">
+      <button id="fileSelect">Load Bug Report</button>
```
Comment (piotrswigon) [inline devtools/src/arcs-bug-report.js]: This UI will only be visible when Arcs Explorer is already connected to some Arcs Runtime, otherwise if Arcs Explorer is not connected to anything you will see something like:

Screen Shot 2019-09-24 at 10 47 31 am

(this is rendered by arcs-connection-status FWIW)

The right thing to do would be to upload the Bug Report _instead_ of having a live connection, so I would imagine having something like:

Screen Shot 2019-09-24 at 10 50 27 am

Now if you upload the bug report this Arcs Explorer instance would no longer wait for Arcs Shell, actually it should disconnect from the ALDS entirely, so that another Arcs Explorer can be used to exploration of the live shell.

Another option that may be easier is to have a URL parameter `?bugreport` that is handled like other connection types in devtools.js and will cause opening up a file picker instead of the websocket or webrtc connection. This "Load a Bug Report" link could just link to "href=?bugreport", then it would all just work without much work I think...

WDYT?

**Comment by piotrswigon** [comment]
On  a second thought. Maybe we should call it a "snapshot" instead of a "bug report" ? WDYT? We're not filing a bug report, at least not yet, so "snapshot" feels more like the right word.

We could use an 'image:photo-camera' icon to have that icon:
Screen Shot 2019-09-30 at 8 00 51 am

with maybe "Take a snapshot" as a title text. And then say "Load from a snapshot"...

Let me know what you think.

File: devtools/src/arcs-bug-report.js (hunk 1 above)
Comment (yuangu-google) [reply devtools/src/arcs-bug-report.js]: Adding that option fixes the issue.

File: devtools/src/arcs-bug-report.js (hunk 2 above)
Comment (yuangu-google) [reply devtools/src/arcs-bug-report.js]: Done.

File: devtools/src/arcs-bug-report.js (hunk 3 above)
Comment (yuangu-google) [reply devtools/src/arcs-bug-report.js]: I added one icon to the left of these two buttons, which should make it clear that this section is bug related...

File: devtools/src/arcs-bug-report.js (hunk 6 above)
Comment (yuangu-google) [reply devtools/src/arcs-bug-report.js]: Adding the parameter sgtm.

**Comment by yuangu-google** [comment]
I changed from bugreport to snapshot which seems a clearer name to me as well.

**Comment by piotrswigon** [comment]
Question: Would you like to land your change and be done with it, or would you be ok with some more tinkering?

The feature essentially works. There are some minor issues (e.g. it turns out you cannot open a file picker without user click), but overall we're good. I don't want to frustrate you with a lot of back-and-forth, so let's agree on the extent of your appetite for hacking on this :)

I would however like to make some UI changes. In particular: I would like the camera icon to be the button to perform the snapshot, I would like it to be shaded like other such buttons, which requires some CSS changes, etc.

I would also like to structure things a bit differently. There is this "connection" abstraction at the bottom of the webapp. I have plans for some future improvements, e.g. for the shell to say what are its capabilities, so that Arcs Explorer can hide/show certain tools, e.g. Strategizer and Planner tabs don't make sense if the given shell is not doing planning (e.g. dev-shell and pipes-shell don't plan). These future changes are not in scope for your feature - I'm just describing what I'm planning for future.

So I would like the replay of a snapshot to become a new type of "connection", and then in the future it could introduce itself as a kind of "readonly shell", which means all the feature that require request-response calls (e.g. recipe editor) will be disabled or hidden.

Given that the file picker needs to be opened with a click, I think the cleanest way out is to ask the user what kind of connection they want: WebSocket, WebRTC or Snapshot. Unless they've put certain URL params, in which case we would just load the request connection. I would envision that when you visit devtools webapp you would get a modal dialog asking about this, something like this:

Screen Shot 2019-10-02 at 3 14 57 pm

We can do 2 things: Either we make minimal changes to your PR and land it, and then I take care of these various things I care about. Or if you have an appetite for working with me on these things, you can own implementing these changes.

Let me know.
